jax/experimental/pjit.py

```diff
@@ -322,6 +325,31 @@ def _pjit_init_to_final_params(params):
   return bind_params
 core.initial_to_final_param_rules[pjit_call_p] = _pjit_init_to_final_params
 
+def _check_resources_against_named_axes(what, aval, pos_axis_resources, named_axis_resources):
+  pjit_resources = set(it.chain.from_iterable(pos_axis_resources))
+  aval_resources = set(it.chain.from_iterable(
+    named_axis_resources[a] for a in aval.named_shape))
+  overlap = pjit_resources & aval_resources
+  if overlap:
+    raise JAXTypeError(
```

Comment:
Is this really a TypeError? Seems more like a ValueError (we don't have a JAXValueError yet but presumably we could?). Or if you're fancy, you could add a new JAXMeshAxisAlreadyInUse or something.

Reply:
Yeah we should add custom errors for this. For now I treat them as "Jaxpr type errors", so JAXTypeError seemed fitting 🤷